viewlog.c

```diff
@@ -91,34 +101,54 @@ void ViewLog(int config)
   CloseHandle(proc_info.hProcess);
 }
 
-
-void EditConfig(int config)
+/* either config or config_path */
+void EditConfig(int config, PTCHAR config_path)
 {
-  TCHAR filename[2*MAX_PATH];
+  TCHAR config_file[2*MAX_PATH];
+  TCHAR fullpath[2*MAX_PATH];
+  TCHAR config_dir[2*MAX_PATH];
+  TCHAR assocexe[2*MAX_PATH];
+  DWORD assocexe_num;
 
   STARTUPINFO start_info;
   PROCESS_INFORMATION proc_info;
   SECURITY_ATTRIBUTES sa;
   SECURITY_DESCRIPTOR sd;
-  HINSTANCE status;
+  HINSTANCE status = 0;
 
   CLEAR (start_info);
   CLEAR (proc_info);
   CLEAR (sa);
   CLEAR (sd);
 
-  /* Try first using file association */
-  _sntprintf_0(filename, L"%s\\%s", o.conn[config].config_dir, o.conn[config].config_file);
+  if (config_path == NULL)
+  {
+    // using "config"
+    _sntprintf_0(fullpath, L"%s\\%s", o.conn[config].config_dir, o.conn[config].config_file);
+    _tcsncpy(config_dir, o.conn[config].config_dir, _countof(config_dir) - 1);
+    _tcsncpy(config_file, o.conn[config].config_file, _countof(config_file) - 1);
+  }
+  else
+  {
+    // using "config_path"
+    _tcsncpy(fullpath, config_path, _countof(config_dir) - 1);
+    _tcsncpy(config_dir, config_path, _countof(config_dir) - 1);
+    PathRemoveFileSpec(config_dir);
+    _tcsncpy(config_file, config_path, _countof(config_file) - 1);
+    *config_file = PathFindFileName(config_file);
+  }
 
+  /* Try first using file association */
   CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE); /* Safe to init COM multiple times */
-  status = ShellExecuteW (o.hWnd, L"open", filename, NULL, o.conn[config].config_dir, SW_SHOWNORMAL);
+  if (AssocQueryString(0, ASSOCSTR_EXECUTABLE, L".txt", NULL, assocexe, &assocexe_num) == S_OK)
+    status = ShellExecuteW (o.hWnd, L"open", assocexe, fullpath, config_dir, SW_SHOWNORMAL);
   if (status > (HINSTANCE) 32)
     return;
   else
     PrintDebug (L"Opening config file using ShellExecute with verb = open failed"
-                 " for config '%s' (status = %lu)", o.conn[config].config_name, status);
+                 " for config '%s' (status = %lu)", config_file, status);
 
-  _sntprintf_0(filename, _T("%s \"%s\\%s\""), o.editor, o.conn[config].config_dir, o.conn[config].config_file);
+  _sntprintf_0(fullpath, _T("%s \"%s\""), o.editor, fullpath);
```

Comment:
Overlapped destination and argument string in snprintf family is undefined.  And this particular usage is guaranteed to mess up fullpath.